core/chains/evm/txm/storage/inmemory_store.go

```diff
@@ -0,0 +1,358 @@
+			delete(m.UnconfirmedTransactions, *tx.Nonce)
+		}
+	}
+
+	var unconfirmedTransactionIDs []uint64
+	for _, tx := range m.ConfirmedTransactions {
+		if tx.Nonce == nil {
+			return nil, nil, fmt.Errorf("nonce for txID: %v is empty", tx.ID)
+		}
+		existingTx, exists := m.UnconfirmedTransactions[*tx.Nonce]
+		if exists {
+			m.lggr.Errorw("Another unconfirmed transaction with the same nonce exists. Transaction will overwritten.",
+				"existingTx", existingTx, "newTx", tx)
+		}
+		if *tx.Nonce >= latestNonce {
+			tx.State = txmgr.TxUnconfirmed
+			tx.LastBroadcastAt = nil // Mark reorged transaction as if it wasn't broadcasted before
+			unconfirmedTransactionIDs = append(unconfirmedTransactionIDs, tx.ID)
+			m.UnconfirmedTransactions[*tx.Nonce] = tx
+			delete(m.ConfirmedTransactions, *tx.Nonce)
+		}
+	}
+
+	if len(m.ConfirmedTransactions) > maxQueuedTransactions {
+		prunedTxIDs := m.pruneConfirmedTransactions()
+		m.lggr.Debugf("Confirmed transactions map for address: %v reached max limit of: %d. Pruned 1/%d of the oldest confirmed transactions. TxIDs: %v",
+			m.address, maxQueuedTransactions, pruneSubset, prunedTxIDs)
+	}
+	sort.Slice(confirmedTransactions, func(i, j int) bool { return confirmedTransactions[i].ID < confirmedTransactions[j].ID })
+	sort.Slice(unconfirmedTransactionIDs, func(i, j int) bool { return unconfirmedTransactionIDs[i] < unconfirmedTransactionIDs[j] })
+	return confirmedTransactions, unconfirmedTransactionIDs, nil
+}
+
+func (m *InMemoryStore) MarkUnconfirmedTransactionPurgeable(nonce uint64) error {
+	m.Lock()
+	defer m.Unlock()
+
+	tx, exists := m.UnconfirmedTransactions[nonce]
+	if !exists {
+		return fmt.Errorf("unconfirmed tx with nonce: %d was not found", nonce)
+	}
+
+	tx.IsPurgeable = true
+
+	return nil
+}
+
+func (m *InMemoryStore) UpdateTransactionBroadcast(txID uint64, txNonce uint64, attemptHash common.Hash) error {
+	m.Lock()
+	defer m.Unlock()
+
+	unconfirmedTx, exists := m.UnconfirmedTransactions[txNonce]
+	if !exists {
+		return fmt.Errorf("unconfirmed tx was not found for nonce: %d - txID: %v", txNonce, txID)
+	}
+
+	// Set the same time for both the tx and its attempt
+	now := time.Now()
+	unconfirmedTx.LastBroadcastAt = &now
+	if unconfirmedTx.InitialBroadcastAt == nil {
+		unconfirmedTx.InitialBroadcastAt = &now
+	}
+	a, err := unconfirmedTx.FindAttemptByHash(attemptHash)
+	if err != nil {
+		return fmt.Errorf("UpdateTransactionBroadcast failed to find attempt. %w", err)
+	}
+	a.BroadcastAt = &now
+
+	return nil
+}
+
+func (m *InMemoryStore) UpdateUnstartedTransactionWithNonce(nonce uint64) (*types.Transaction, error) {
+	m.Lock()
+	defer m.Unlock()
+
+	if len(m.UnstartedTransactions) == 0 {
+		m.lggr.Debugf("Unstarted transactions queue is empty for address: %v", m.address)
+		return nil, nil
+	}
+
+	if _, exists := m.UnconfirmedTransactions[nonce]; exists {
+		return nil, fmt.Errorf("an unconfirmed tx with the same nonce already exists: %v", m.UnconfirmedTransactions[nonce])
+	}
```

Comment:
nit: think we can avoid 2 map accesses here
```suggestion
	if tx, exists := m.UnconfirmedTransactions[nonce]; exists {
		return nil, fmt.Errorf("an unconfirmed tx with the same nonce already exists: %v", tx)
	}
```